Improve MeshPhysicalMaterial transmission support Part2

takahirox opened the pull request:
Related issue: https://github.com/mrdoob/three.js/pull/21884#issuecomment-858649794

**Description**

This PR improves `MeshPhysicalMaterial` `transmission` support.

![image](https://user-images.githubusercontent.com/7637832/121625624-ade68380-ca28-11eb-946d-45d3d8ea1135.png)

@WestLangley Looking good now?

**Changes**

1. In the previous PR I first applied transmission to `diffuseColor`

https://github.com/mrdoob/three.js/blob/r129/src/renderers/shaders/ShaderChunk/transmission_fragment.glsl.js#L30

and then applied lighting to it. This is the main reason why the color was darken even if transmission is 1. In this PR I apply the transmission after the lighting.

2. In the previous PR, I calculated specular in the transmission function. But specular is calculated in the lighting. I speculate I can reuse it.

https://github.com/mrdoob/three.js/blob/r129/src/renderers/shaders/ShaderChunk/transmission_pars_fragment.glsl.js#L80-L82

3. I turn off tone mapping for the first opaque objects pass to avoid to apply it again in the second transmissive objects pass.

**Screenshots**

I still have an issue that multi sample render target is black in the screenshots. So I made the screenshots by temporarily changing the code to non-multi-sample render target. Please feel free to update the screenshots if you can make correctly.

takahirox committed:
Improve MeshPhysicalMaterial transmission

takahirox committed:
Update screenshots

mrdoob commented:
> I still have an issue that multi sample render target is black in the screenshots. So I made the screenshots by temporarily changing the code to non-multi-sample render target. Please feel free to update the screenshots if you can make correctly.

Have you tried running `npm update` inside the `test` folder?

WestLangley commented:
Now that it is looking better, we need to figure out how to properly deal with `premultipliedAlpha` -- when to use it, and when not to.

For a later PR, of course.

WestLangley commented:
Also, `_transmissionRenderTarget` should not be `UnsignedByteType`, because that leads to clamping and texture hue shifts.

Perhaps `UnsignedByteType` can be a fall-back if `HalfFloatType` is not supported.

`UnsignedByteType`
<img width="877" alt="Screen Shot 2021-06-11 at 7 49 42 AM" src="https://user-images.githubusercontent.com/1000017/121682886-a3f56c80-ca8a-11eb-8265-387d9bad02e2.png">


`HalfFloatType` is mostly better :-)
<img width="860" alt="Screen Shot 2021-06-11 at 7 49 59 AM" src="https://user-images.githubusercontent.com/1000017/121683068-ddc67300-ca8a-11eb-820e-19431b686cf1.png">

takahirox committed:
WebGLRenderer: Use HalfFloatType for transmissionRenderTarget

takahirox committed:
Update screenshots

takahirox commented:
@mrdoob 

> Have you tried running `npm update` inside the `test` folder?

Ah, I didn't notice that I needed to run it in the `test` directory. Updated the screenshots, thanks!

@WestLangley 

> Also, `_transmissionRenderTarget` should not be `UnsignedByteType`, because that leads to clamping and texture hue shifts.
>
>Perhaps `UnsignedByteType` can be a fall-back if `HalfFloatType` is not supported.

Good catch. Updated, thanks!

mrdoob commented:
Excellent! Thanks!

mrdoob commented on src/renderers/WebGLRenderer.js:
```diff
@@ -1286,8 +1287,15 @@ function WebGLRenderer( parameters = {} ) {
 
+		// Turn off the features which can affect the frag color for opaque objects pass.
+		// Otherwise they are applied twice in opaque objects pass and transmission objects pass.
+		const currentToneMapping = _this.toneMapping;
+		_this.toneMapping = NoToneMapping;
```
Oh! Actually... I guess we should also set `outputEncoding` to `LinearEncoding`?

takahirox replied:
No because we use renderTarget texture's encoding as output encoding if `currentRenderTarget` isn't `null`.

https://github.com/mrdoob/three.js/blob/r129/src/renderers/webgl/WebGLPrograms.js#L174

If we start to copy the framebuffer we will need to convert from sRGB to linear and from tone mapping to none in the shader if they are set.
